Automattic/wp-calypso
# Media: Manage modal current view using Redux state (2nd Attempt)

This is a second attempt at #8625. After merging, a minor issue was noted with the incorrect prop being called upon component unmounting, fixed in #8818. Shortly thereafter, a more severe issue was reported in #8822, that nothing was shown in the media modal when attempting to set a featured image when not having already toggled the media modal in that same session. An attempt was made to fix this in #8827 by defaulting to the list view. However, an issue remained that when toggling the "Add Media" modal flow, changing the selected view (e.g. Detail), then dismissing and choosing to assign a featured image, the previous view would still be visible. This is because the component never truly unmounts during the same post editor session, only when navigating away from the page. A full revert was made in #8830. The changes in this pull request include all of the above work (#8691, #8818, #8827) plus an additional fix to reset the view when the modal props change to being made not visible (b22a720). There is related logic here to reset state to its initial value when the modal is _shown_. This should probably be made consistent, but I did not want to include additional changes, and it's necessary to reset the view after being made hidden since some flows will set the selected view before showing the modal.

Original pull request text:

-----

Related: #8625

This pull request seeks to migrate the post editor media modal to use Redux state to manage the currently active view of the modal. This will be used as part of the Site Icon project to allow us to control the behavior of the confirmation action, in order to send users through the flow of cropping an image after making a selection. This is also a general refactoring of the modal, relying on global Redux state rather than the modal's component state to maintain the selected view.

__Testing Instructions:__

Verify that there is no change in behavior of using the media modal. In particular, test:

- View media detail
- Insert multiple images as gallery
- View media detail when multiple items selected, and verify behavior of deleting one item
- Use the image editor
- Selecting a gallery and clicking the Edit inline toolbar should launch the modal to the gallery view
- "Back to Media Library"
- Cancelling modal should reset to list next time modal is opened

/cc @gwwar @artpi @lamosty
